packages/angular_devkit/core/src/json/schema/registry.ts

```diff
@@ -175,39 +195,138 @@ export class CoreSchemaRegistry implements SchemaRegistry {
+   * that is also easier to traverse. Does not cache the result.
+   *
+   * @param {JsonObject | string} schema The schema or URI to flatten.
+   * @returns {Observable<JsonObject>} An Observable of the flattened schema object.
+   */
+  flatten(schema: JsonObject | string): Observable<JsonObject> {
+    if (typeof schema == 'string') {
+      return this._fetch(schema).pipe(
+        switchMap(fetchedJson => this.flatten(fetchedJson)),
+      );
+    }
+
+    this._ajv.removeSchema(schema);
+
+    // Supports both synchronous and asynchronous compilation, by trying the synchronous
+    // version first, then if refs are missing this will fails.
+    // We also add any refs from external fetched schemas so that those will also be used
+    // in synchronous (if available).
+    let validator: Observable<ajv.ValidateFunction>;
+    try {
+      this._currentCompilationSchemaInfo = null;
+      validator = of(this._ajv.compile(schema)).pipe(
+        tap(() => this._currentCompilationSchemaInfo = undefined),
+      );
+    } catch (e) {
+      // Propagate the error.
+      if (!(e instanceof (ajv.MissingRefError as {} as Function))) {
+        return throwError(e);
+      }
+
+      this._currentCompilationSchemaInfo = null;
+      validator = from(this._ajv.compileAsync(schema)).pipe(
+        tap(() => this._currentCompilationSchemaInfo = undefined),
+      );
+    }
+
+    return validator.pipe(
+      switchMap(validate => {
+        const self = this;
+
+        function visitor(
+          current: JsonObject | JsonArray,
+          pointer: JsonPointer,
+          parentSchema?: JsonObject | JsonArray,
+          index?: string,
+        ) {
+          if (current
+            && parentSchema
+            && index
+            && isJsonObject(current)
+            && current.hasOwnProperty('$ref')
+            && typeof current['$ref'] == 'string'
+          ) {
+            const resolved = self._resolver(current['$ref'] as string, validate);
+
+            if (resolved.schema) {
+              (parentSchema as JsonObject)[index] = resolved.schema;
+            }
+          }
+        }
+
+        const schema = deepCopy(validate.schema as JsonObject);
+        visitJsonSchema(schema, visitor);
+
+        return of(schema);
+      }),
+    );
   }
 
-  compile(schema: JsonObject): Observable<SchemaValidator> {
-    const schemaKey = serialize(schema);
-    const existingValidator = this._validatorCache.get(schemaKey);
-    if (existingValidator) {
-      return observableOf(existingValidator);
+  /**
+   * Compile and return a validation function for the Schema.
+   *
+   * @param {JsonObject | string} schema The schema to validate. If a string, will fetch the schema
+   *     before compiling it (using schema as a URI).
+   * @returns {Observable<SchemaValidator>} An Observable of the Validation function.
+   */
+  compile(schema: JsonObject | string): Observable<SchemaValidator> {
```

Comment:
Is the string variant used anywhere?  Also, I think that by looking at the signature, I would assume that the string would be the actual JSON content rather than a URI.

Reply:
No, I'll remove.

Reply:
I can see the concept being useful, however.  Maybe a separate method?

Reply:
Out of scope of this PR, but yes.